pandas/core/frame.py

```diff
@@ -2634,6 +2634,52 @@ def transpose(self, *args, **kwargs):
 
     T = property(transpose)
 
+    # ----------------------------------------------------------------------
+    # Array Interface
+
+    # This is also set in IndexOpsMixin
+    # GH#23114 Ensure ndarray.__op__(DataFrame) returns NotImplemented
+    __array_priority__ = 1000
+
+    def __array__(self, dtype=None):
+        return com.values_from_object(self)
+
+    def __array_wrap__(self, result: np.ndarray,
```

Comment:
Hmm, I think this can be cleaned up. Why don't we apply the ufunc on each block and combine the results? 

IMO, `np.sin(df)` should essentially be

1. `arrays = [np.sin(block) for block in df._data]
2. `pd.DataFrame._from_arrays(arrays, df.columns, df.index)`

i.e. call `sin` on each block & reconstruct the result.

Reply:
in theory this should work, but this opens up a huge can of worms as we don't have ``__array_ufunc__`` for series or frame, so all of the ops need work. This would need to be another issue.